src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/PinnedGCHandle.T.cs

```diff
@@ -0,0 +1,122 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// Represents an strongly-typed opaque, GC handle to a managed object.
+    /// A GC handle is used when an object reference must be reachable from
+    /// unmanaged memory.
+    /// The object is pinned at fixed location in GC heap and allows its
+    /// address to be taken.
+    /// </summary>
+    /// <remarks>
+    /// <see cref="PinnedGCHandle{T}"/> corresponds to Pinned roots.
+    /// For Normal, see <see cref="GCHandle{T}"/>.
+    /// For Weak and WeakTrackResurrection, see <see cref="WeakGCHandle{T}"/>.
+    /// </remarks>
+    /// <seealso cref="GCHandle" />
+    /// <typeparam name="T">The type of the object this <see cref="GCHandle{T}"/> tracks to.</typeparam>
+    public struct PinnedGCHandle<T> : IDisposable
+        where T : class?
+    {
+        // The actual integer handle value that the EE uses internally.
+        private IntPtr _handle;
+
+        /// <summary>
+        /// Allocates a handle for the specified object.
+        /// </summary>
+        /// <param name="target">The object that uses the <see cref="GCHandle{T}"/>.</param>
+        public PinnedGCHandle(T target)
+        {
+            GCHandle.ThrowIfNotPinnable(target);
```

Comment:
This check should not be present here. Pinning any type is allowed with this new type. The restriction on the old GC handle was just for historical reasons. The runtime already supports pinning any kind of object.